### ISSUE TITLE
Query time problem when filtering by rating and year.

### ISSUE THREAD

**Issue by hakanersu** [hakanersu/iwatched#2]
It seems when filtering with year and rating, query time takes nearly 1-3 seconds.

```sql
select * from "titles" where exists (select * from "ratings" where "titles"."tconst" = "ratings"."tconst" and ("average_rating" > 8 and "average_rating" < 9)) and "title_type" = 'movie' and "start_year" = '2007' order by "weight" desc limit 11 offset 0
```

After removing `order by` query time down to 0.300ms but i really want to keep weight order by. And removing where year also improves performance of the query.

![image](https://user-images.githubusercontent.com/1052467/84580282-20a12c80-adde-11ea-84ba-ff7f5eeb1562.png)

**Comment by ordago** [comment]
Hey @hakanersu , ahve you seen this talk?

[Things every developer absolutely, positively needs to know about database indexing - Kai Sassnowski](https://www.youtube.com/watch?v=HubezKbFL7E&)

**Comment by hakanersu** [comment]
@ordago Yes i watched for a while ago and even with indexes still can't figure out the problem yet :smile: I have to learn PostgreSQL explain

**Comment by garyblankenship** [comment]
`CREATE INDEX "ratings_average" ON "public"."ratings" USING BTREE ("average_rating");`
`CREATE INDEX "titles_type_index" ON "public"."titles" USING BTREE ("title_type");`
`CREATE INDEX "titles_start_year_index" ON "public"."titles" USING BTREE ("start_year");`

added indexes on 3 columns locally and the query returns for me in 43ms.  I'm not very famliar with postgres either and used TablePlus to add these.

**Comment by ordago** [comment]
`("average_rating" > 8 and "average_rating" < 9)`

This wouldn't include  an average rating of exactly 8. 

In ProgressSQL you can have an index on the result of a function.

https://www.postgresql.org/docs/9.1/indexes-expressional.html

You could make an index on `floor(average_rating)`, that way you could do `(floor(average_rating) = 8)` and it would use the index.


Also, have you tried changing the order of the columns in the where clause?

**Comment by hakanersu** [comment]
@garyblankenship First thank you, I added that indexes too but query performance is same for me. Here is the results with your given indexes. But i'll try it with a clean setup when i have time.
![image](https://user-images.githubusercontent.com/1052467/84595376-39eabd00-ae60-11ea-8580-d4fbf6236fc4.png)

@ordago Yes you are right about including rating 8 and your solution about `(floor(average_rating) = 8)` is way better that current solution. I tried to adding floor(avarage_rating) index but it didn't work for now but i'll digg deeper about it.

By the way i think real problem is `weight` column. When i remove `weight order by` query time down from 3-4sec to 200ms. But in only occurs when adding rating filter. 

For example when i filter year and watched status query time is only 6.2ms 

```sql
select * from "titles" where "title_type" = 'movie' and not exists (select * from "watched" where "titles"."tconst" = "watched"."tconst" and "user_id" = 1) and "start_year" = '2014' order by "weight" desc limit 11 offset 0
6.26ms
```

**Comment by hakanersu** [comment]
@garyblankenship Can you share your all indexes, i tested and its works when i remove tconst index.

**Comment by garyblankenship** [comment]
| tablename | indexname               | indexdef                                                                       |
|-----------|-------------------------|--------------------------------------------------------------------------------|
| titles    | titles_pkey             | CREATE UNIQUE INDEX titles_pkey ON public.titles USING btree (id)              |
| titles    | titles_start_year_index | CREATE INDEX titles_start_year_index ON public.titles USING btree (start_year) |
| titles    | titles_tconst_index     | CREATE INDEX titles_tconst_index ON public.titles USING btree (tconst)         |
| titles    | titles_type_index       | CREATE INDEX titles_type_index ON public.titles USING btree (title_type)       |
| titles    | titles_weight_index     | CREATE INDEX titles_weight_index ON public.titles USING btree (weight)         |
